### PR TITLE
Oracle review 2

### PR DESCRIPTION
**Pull request by btrif**
I made the requested changes.
There would be to implement logging, but unfortunately my time is limited.

### CODE REVIEW

File: routes.py [hunk 1]
```diff
@@ -28,36 +28,41 @@ def list_cars():
     if cars:
         return jsonify({'cars': list_of_cars})
```
Comment (kingdomik) [inline routes.py]: As I see sometimes service returns "{'status': 'success', ..." but not here, let's use standard approach everywhere

File: routes.py [hunk 2]
```diff
@@ -28,36 +28,41 @@ def list_cars():
...
 def add_car():  # in browser URL : http://127.0.0.1:7000/add?car_number=CJ45WAY&tariff=hourly
 
-    car_number = request.args.get('car_number')
+    new_car_number = request.args.get('car_number')
     tariff = request.args.get('tariff')
 
+    # Check if there is no car in the parking with the same car_number:
+    existing_car_number_count = Parking.query.filter_by(car_number=new_car_number).count()
+    if existing_car_number_count > 0:
+        return {'status': 'error', 'message': 'Hey! There is already a car with that number. I will call the Police !'}
```
Comment (kingdomik) [inline routes.py]: The error message should help user to identify and fix the issue but seems customers getting such errors will call our support always.

File: routes.py [hunk 3]
```diff
@@ -28,36 +28,41 @@ def list_cars():
...
+
     # Check that the tariff is only hourly or daily :
     if tariff not in ['hourly', 'daily']:
-        raise NameError('the tariff must be either hourly or daily')
+        return {'status': 'error', 'message': 'the tariff must be either hourly or daily'}
```
Comment (kingdomik) [inline routes.py]: Error message should provide complete information including incorrect value

File: routes.py [hunk 4]
```diff
@@ -28,36 +28,41 @@ def list_cars():
...
         try:
             db.session.add(new_car)
             db.session.commit()
-            Settings.available_places -= 1      # Update available places
-            return {'status': 'success', 'id': new_car.id, 'car_number': new_car.car_number,
+            Settings.available_places -= 1  # Update available places
+            return {'status': 'success', 'id': new_car.id, 'new_car_number': new_car.car_number,
                     'tariff': new_car.tariff, 'date_start': new_car.date_start,
                     'available_places': Settings.available_places}
         except Exception:
```
Comment (kingdomik) [inline routes.py:80]: 'there was an error adding your car' how to investigate such failures? Let's make it clear to don't overload our support with customer's calls

Comment (kingdomik) [reply routes.py:80]: Also, shouldn't result be like "{'status': 'error',...}"?

**Review by kingdomik — COMMENTED** [review]
Conceptually I don't see a big difference with the previous pull request

**Comment by btrif** [comment]
Hi @kingdomik & @nexdevs .
Thank you for your observations. Fix those but not committed yet.
What It remains to address is the possibility to change the tariffs at runtime.
I'll fix that also and come back.

**Comment by btrif** [comment]
@nexdevs @kingdomik 

I implemented the runtime scenario where you can change tariffs and parking capacity


Bogdan

File: utilities.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    :return: int, the Fee    '''
+    from math import ceil
+    if seconds <= 900:  # 15 min * 60 secs = 900 secs
+        return 0
+    elif tariff_type == 'hourly':
+        return ceil(seconds / 3600) * tariff_value
+    elif tariff_type == 'daily':
+        return ceil(seconds / 86400) * tariff_value
+    else:
+        return f'Something got wrong. Please check'
```
Comment (kingdomik) [inline utilities.py:28]: As we discussed by zoom it should be possible to easily add tariffs with new LOGIC but it is hardcoded still.

File: routes.py [hunk 6]
```diff
@@ -26,66 +31,111 @@ def list_cars():
...
+            config['parking_info']['hourly_tariff'] = hourly_tariff
+            config['parking_info']['daily_tariff'] = daily_tariff
+            config['parking_info']['parking_capacity'] = parking_capacity
+            PC.write_parking_configuration()
+            return {"status": "success",
+                    "message": f"hourly_tariff tariff was updated to {hourly_tariff}, "
+                               f" daily tariff was updated to {daily_tariff}"
+                               f" parking capacity was updated to {parking_capacity}"}
+        else:
+            return {"status": "error", "message": f"You must input integer or float"}
```
Comment (kingdomik) [inline routes.py:141]: Poor error reporting. The error message doesn't allow clients to easily identify and fix the issue.

File: routes.py [hunk 7]
```diff
@@ -26,66 +31,111 @@ def list_cars():
...
+                                                     'tariff': car.tariff,
+                                                     'date_start': car.date_start,
+                                                     'date_end': datetime.utcnow(),
+                                                     'fee': fee,
+                                                     'available_places': available_places,
                                                      'time_diff': str(date_end - car.date_start), 'seconds': seconds}
                     }
+
         except Exception:
             return {'status': 'error', 'message': f"there was an error removing your car with id {car.id}"}
```
Comment (kingdomik) [inline routes.py:113]: Poor error reporting. The error message doesn't allow clients to easily identify and fix the issue.

File: routes.py [hunk 8]
```diff
@@ -26,66 +31,111 @@ def list_cars():
...
-            Settings.available_places -= 1      # Update available places
-            return {'status': 'success', 'id': new_car.id, 'car_number': new_car.car_number,
+            available_places -= 1  # Update available places
+            return {'status': 'success', 'id': new_car.id, 'new_car_number': new_car.car_number,
                     'tariff': new_car.tariff, 'date_start': new_car.date_start,
-                    'available_places': Settings.available_places}
+                    'available_places': available_places}
         except Exception:
-            return 'there was an error adding your car'
+            return {'status': 'error', 'message': 'The car was not added. Connection problem with the database.'}
```
Comment (kingdomik) [inline routes.py:81]: Poor error reporting. The error message doesn't allow clients to easily identify and fix the issue.

File: routes.py [hunk 9]
```diff
@@ -26,66 +31,111 @@ def list_cars():
...
+        date_start = Parking.query.filter_by(car_number='CT33M').all()[0].date_start
+        position = Parking.query.filter_by(car_number='CT33M').all()[0].id
+        return {'status': 'error', 'message': f"There is already a parked car with this number since {date_start}"
+                                              f". It can be found at the position {position}"}
+
     # Check that the tariff is only hourly or daily :
     if tariff not in ['hourly', 'daily']:
-        raise NameError('the tariff must be either hourly or daily')
+        return {'status': 'error',
+                'message': f"{tariff} is not a Tariff Plan. tariff must be either `hourly` or `daily`"}
```
Comment (kingdomik) [inline routes.py:55]: Hardcoded tariff names while expected more flexibility

**Review by kingdomik — COMMENTED** [review]
@btrif great thanks for improving the solution, I see the following issues
- Tariffs are still hardcoded while would be flexible (discussed by zoom like free parking on holidays, etc.)
- Error reporting is still not clear in some places (see comments). Error reporting doesn't cover general code mistakes but only specific issues + resource not found
- No logging while requested in PR before and by zoom

File: utilities.py (hunk 5 above)
Comment (btrif) [reply utilities.py:28]: @kingdomik Tariffs are not hardcoded. Please see that I added a section called /configuration where you can change the tariffs of the parking and also the parking capacity (supposing that the app will be handled to more parkings )

image

Comment (btrif) [reply utilities.py:28]: @kingdomik  The calculation algorithm of course is hardcoded. How this cannot be hardcoded ?

Comment (btrif) [reply utilities.py:28]: @kingdomik  Can you put the new requirements into a new document and send me by e-mail ?
This way I'll have everything organized in one place.

Comment (kingdomik) [reply utilities.py:28]: > How this cannot be hardcoded ?

Hi @btrif 

> @kingdomik Tariffs are not hardcoded. Please see that I added a section called /configuration where you can change the tariffs of the parking and also the parking capacity (supposing that the app will be handled to more parking )

Thanks for making tariffs parameters configurable

> @kingdomik The calculation algorithm of course is hardcoded. How this cannot be hardcoded ?

Could you please refactor the code to create/delete/update tariff plans without the core code changes

> @kingdomik Can you put the new requirements into a new document and send me by e-mail ?
This way I'll have everything organized in one place.

Let me continue in the pull request, it contains whole our discussion

**Pull request closed without merging**